pkg/controller/v1alpha1/llmisvc/router_discovery.go

```diff
@@ -0,0 +1,203 @@
+		if err := c.Get(ctx, types.NamespacedName{Name: string(gateway.Spec.GatewayClassName)}, gatewayClass); err != nil {
+			return nil, fmt.Errorf("failed to get GatewayClass %q for gateway %s/%s: %w", string(gateway.Spec.GatewayClassName), gwNS, gwName, err)
+		}
+		gateways = append(gateways, resolvedGateway{
+			gateway:      gateway,
+			gatewayClass: gatewayClass,
+			parentRef:    parentRef,
+		})
+	}
+	return gateways, nil
+}
+
+func DiscoverURLs(ctx context.Context, c client.Client, route *gwapiv1.HTTPRoute) ([]*apis.URL, error) {
+	var urls []*apis.URL
+
+	gateways, err := discoverGateways(ctx, c, route)
+	if err != nil {
+		return nil, fmt.Errorf("failed to discover gateways: %w", err)
+	}
+
+	for _, g := range gateways {
+		listener := selectListener(g.gateway, g.parentRef.SectionName)
+		scheme := extractSchemeFromListener(listener)
+		port := listener.Port
+
+		addresses := g.gateway.Status.Addresses
+		if len(addresses) == 0 {
+			return nil, &ExternalAddressNotFoundError{
+				GatewayNamespace: g.gateway.Namespace,
+				GatewayName:      g.gateway.Name,
+			}
+		}
+
+		hostnames := extractRouteHostnames(route)
+		if len(hostnames) == 0 {
+			hostnames = extractAddressValues(addresses)
+		}
+
+		path := extractRoutePath(route)
+
+		gatewayURLs, err := combineIntoURLs(hostnames, scheme, port, path)
+		if err != nil {
+			return nil, fmt.Errorf("failed to combine URLs for Gateway %s/%s: %w", g.gateway.Namespace, g.gateway.Name, err)
+		}
+
+		urls = append(urls, gatewayURLs...)
+	}
+
+	return urls, nil
+}
+
+func extractRoutePath(route *gwapiv1.HTTPRoute) string {
+	if len(route.Spec.Rules) > 0 && len(route.Spec.Rules[0].Matches) > 0 {
+		// TODO how do we deal with regexp
+		return ptr.Deref(route.Spec.Rules[0].Matches[0].Path.Value, "/")
+	}
+	return "/"
+}
+
+func selectListener(gateway *gwapiv1.Gateway, sectionName *gwapiv1.SectionName) *gwapiv1.Listener {
+	if sectionName != nil {
+		for _, listener := range gateway.Spec.Listeners {
+			if listener.Name == *sectionName {
+				return &listener
+			}
+		}
+	}
+
+	return &gateway.Spec.Listeners[0]
+}
+
+func extractSchemeFromListener(listener *gwapiv1.Listener) string {
+	if listener.Protocol == gwapiv1.HTTPSProtocolType {
+		return "https"
+	}
+	return "http"
+}
+
+func extractRouteHostnames(route *gwapiv1.HTTPRoute) []string {
+	var hostnames []string
+	for _, h := range route.Spec.Hostnames {
+		host := string(h)
+		if host != "" && host != "*" {
+			hostnames = append(hostnames, host)
+		}
+	}
+	return hostnames
+}
+
+func extractAddressValues(addresses []gwapiv1.GatewayStatusAddress) []string {
+	var values []string
+	for _, addr := range addresses {
+		if addr.Value != "" {
+			values = append(values, addr.Value)
+		}
+	}
+	return values
+}
+
+func combineIntoURLs(hostnames []string, scheme string, port gwapiv1.PortNumber, path string) ([]*apis.URL, error) {
+	urls := make([]*apis.URL, 0, len(hostnames))
+
+	sortedHostnames := make([]string, len(hostnames))
```

Comment:
Should we explicitly configure the primary host instead of relying on the sorting ?

Reply:
We can address it as a follow-up. The idea is that this function sorts all addresses to have them stable. Then we find first external address and expose is as `.status.url`, everything goes into `.status.addresses`. Is it fine for you @yuzisun ?

https://github.com/kserve/kserve/blob/b5e3d533eb7e2f5b6a02b9d4c01a6baa00fc0eb0/pkg/controller/v1alpha1/llmisvc/router.go#L165-L182